man/mflow.1
.Dd September 6, 2017
.Dt MFLOW 1
.Os
.Sh NAME
.Nm mflow
.Nd reflow format=flowed plain text messages
.Sh SYNOPSIS
.Nm
.Op Fl f
.Op Fl q
.Op Fl w Ar width
\&<
.Ar file
.Sh DESCRIPTION
.Nm
reformats the standard input according to the rules
of RFC 3676.
.Ev PIPE_CONTENTTYPE
is inspected, making this a suitable filter
for
.Sq text/plain
messages for
.Xr mshow 1 .
Messages not specified as
.Sq format=flowed
are output unchanged.
.Pp
Text is reflowed (where allowed) to
fit the width given in the environment variable
.Ev COLUMNS ,
the terminal width, or 80 characters by default.
.Pp
If defined,
the environment variable
.Ev MAXCOLUMNS
specifies the maximum line length.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl f
Force wrapping of long lines.
.It Fl q
Prefix lines with
.Sq Li \&> .
Can be used multiple times.
.It Fl w Ar width
Set maximum line length to
.Ar width .
.El
.Sh EXIT STATUS
.Ex -std
.Sh SEE ALSO
.Xr mshow 1
.Rs
.%A R. Gellens
.%D February 2004
.%R RFC 3676
.%T The Text/Plain Format and DelSp Parameters
.Re
.Sh AUTHORS
.An Leah Neukirchen Aq Mt leah@vuxu.org
.Sh LICENSE
.Nm
is in the public domain.
.Pp
To the extent possible under law,
the creator of this work
has waived all copyright and related or
neighboring rights to this work.
.Pp
.Lk http://creativecommons.org/publicdomain/zero/1.0/


NEWS.md
## 0.3.1 (2018-01-30)

* mless: support $NO_COLOR
* mcolor: support $NO_COLOR
* blaze822.h: ensure PATH_MAX is defined
* Improved documentation.
* Many fixes for address parser.

## 0.3 (2018-01-12)

* New tool mflow to reformat format=flowed plain text mails.
* New tool mbnc to bounce mails (send original mail to someone else).
* mshow filters can output raw text now, e.g. for HTML rendering with colors.
* mrep can quote mail that doesn't have a plain text part.
* mcom runs mmime when deemed necessary.
* mhdr can extract MIME parameters.
* New contrib: mrecode
* New contrib: mraw
* mshow regards non-MIME mails as MIME mails with one part now.
* mshow -F to disable MIME filters.
* mpick supports negations now.
* msed can remove headers depending on their value.
* Improved UTF-8 parsing.
* Improved documentation.
* Numerous bug fixes and portability fixes.

## 0.2 (2017-07-17)

* New sequence syntax `m:+n` for `n` messages after message `m`.
* Threading shortcuts `=`, `_`, `^` for `.=`, `._`, `.^`.
* Sequence related errors are now reported.
* minc and mlist normalize slashes in paths.
* mfwd now generates conforming message/rfc822 parts.
* mthread can add optional folders (e.g. your outbox) to resolve message ids.
* mcom now adds Date: just before sending or cancelling the mail.
* VIOLATIONS.md documents how mblaze works with certain common mistakes.
* Full documentation revamp by Larry Hynes.
* Fix rare crash looking for mail body.
* Numerous small bug and portability fixes.

## 0.1 (2017-06-24)

* Initial release


mflow.c
#include <sys/ioctl.h>
#include <sys/stat.h>

#include <ctype.h>
#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "blaze822.h"

int column = 0;
int maxcolumn = 80;

void
chgquote(int quotes)
{
	static int oquotes;

	if (quotes != oquotes) {
		if (column)
			putchar('\n');
		column = 0;
		oquotes = quotes;
	}
}

void
fixed(int quotes, char *line, size_t linelen)
{
	chgquote(quotes);

	if (column && linelen > (size_t)(maxcolumn - column)) {
		putchar('\n');
		column = 0;
	}

	if (column == 0) {
		for (; column < quotes; column++)
			putchar('>');
		if (quotes)
			putchar(' ');
	}

	fwrite(line, 1, linelen, stdout);
	putchar('\n');
	column = 0;
}

void
flowed(int quotes, char *line, ssize_t linelen)
{
	chgquote(quotes);
	int done = 0;

	while (!done) {
		if (column == 0) {
			for (; column < quotes; column++)
				putchar('>');
			column++;
			if (quotes)
				putchar(' ');
		}

		char *eow;
		if (*line == ' ')
			eow = memchr(line + 1, ' ', linelen - 1);
		else
			eow = memchr(line, ' ', linelen);

		if (!eow) {
			eow = line + linelen;
			done = 1;
		}

		if (column + (eow - line) > maxcolumn &&
		    eow - line < maxcolumn &&
		    column - quotes > 1) {
			putchar('\n');
			column = 0;
			done = 0;
			if (*line == ' ') {
				line++;
				linelen--;
			}
		} else {
			fwrite(line, 1, eow - line, stdout);
			column += eow - line;
			linelen -= eow - line;
			line = eow;
		}
	}
}

int
main(int argc, char *argv[])
{
	char *linebuf = 0;
	char *line;
	size_t linelen = 0;
	int outer_quotes = 0;
	int quotes;

	int reflow = 1;  // re-evaluated on $PIPE_CONTENTTYPE
	int force = 0;
	int delsp = 0;

	char *ct = getenv("PIPE_CONTENTTYPE");
	if (ct) {
		char *s, *se;
		reflow = 0;
		if (blaze822_mime_parameter(ct, "format", &s, &se) && s)
			reflow = (strncasecmp(s, "flowed", 6) == 0);
		if (blaze822_mime_parameter(ct, "delsp", &s, &se) && s)
			delsp = (strncasecmp(s, "yes", 3) == 0);
	}

	char *cols = getenv("COLUMNS");
	if (cols && isdigit(*cols)) {
		maxcolumn = atoi(cols);
	} else {
		struct winsize w;
		int fd = open("/dev/tty", O_RDONLY | O_NOCTTY);
		if (fd >= 0) {
			if (ioctl(fd, TIOCGWINSZ, &w) == 0)
				maxcolumn = w.ws_col;
			close(fd);
		}
	}

	char *maxcols = getenv("MAXCOLUMNS");
	if (maxcols && isdigit(*maxcols)) {
		int m = atoi(maxcols);
		if (maxcolumn > m)
			maxcolumn = m;
	}

	int c;
	while ((c = getopt(argc, argv, "fqw:")) != -1)
		switch (c) {
		case 'f': force = 1; break;
		case 'q': outer_quotes++; break;
		case 'w': maxcolumn = atoi(optarg); break;
		default:
			fprintf(stderr, "Usage: mflow [-f] [-q] [-w MAXCOLUMNS]\n");
			exit(2);
		}

	while (1) {
		errno = 0;
		ssize_t rd = getdelim(&linebuf, &linelen, '\n', stdin);
		if (rd == -1) {
			if (errno == 0)
				break;
			fprintf(stderr, "mflow: error reading: %s\n",
			    strerror(errno));
			exit(1);
		}

		line = linebuf;

		if (!reflow && !force) {
			fwrite(line, 1, rd, stdout);
			continue;
		}

		if (rd > 0 && line[rd-1] == '\n')
			line[--rd] = 0;
		if (rd > 0 && line[rd-1] == '\r')
			line[--rd] = 0;

		quotes = outer_quotes;
		while (*line == '>') {  // measure quote depth
			line++;
			quotes++;
			rd--;
		}

		if (reflow && *line == ' ') {  // space stuffing
			line++;
			rd--;
		}

		if (strcmp(line, "-- ") == 0) {  // usenet signature convention
			if (column)
				fixed(quotes, "", 0);  // flush paragraph
			fixed(quotes, line, rd);
			continue;
		}

		if (reflow && rd > 0 && line[rd-1] == ' ') {  // flowed line
			if (delsp)
				line[--rd] = 0;
			flowed(quotes, line, rd);
		} else {
			if (force && rd > maxcolumn) {
				flowed(quotes, line, rd);
				fixed(quotes, "", 0);
			} else {
				fixed(quotes, line, rd);
			}
		}
	}

	if (reflow && column != 0)
		putchar('\n');
}
